File: 07 - TypeScript/hello_world.ts
// Día 7: TypeScript
// Clase en vídeo: https://www.twitch.tv/videos/1841100654?t=00h17m59s

/*
Esto es un cometnario
*/

// Hola mundo

console.log("Hola, TypeScript!")

// Variables

var myString = "Esto es una cadena de texto"
// myString = 5 Error
console.log(myString)

let myString2: string = "Esto es una cadena de texto"
myString2 = "Aquí cambio el valor de la cadena de texto"
myString2 = "6"
console.log(myString2)
console.log(typeof myString2)

let myNumber: number = 7
myNumber = myNumber + 4
console.log(myNumber)
console.log(typeof myNumber)
console.log(myNumber - 1)
console.log(myNumber)

console.log(myString + " " + myNumber)

let myNumber2 = 6.5
console.log(myNumber2)
console.log(typeof myNumber2)

console.log(myNumber + myNumber2)

let myBool: boolean = false
myBool = true
console.log(myBool)
console.log(typeof myBool)

let myUndefined: undefined
// myUndefined = "myUndefined" Error
console.log(myUndefined)

// Constantes

const myConst = "Mi propiedad constante"
// myConst = "Otro valor" Error
console.log(myConst)

// Controles de flujo

if (myNumber == 10 && myString == "Hola") {
    console.log("El valor es 10")
} else if (myNumber == 11 || myString == "Hola") {
    console.log("El valor es 11")
} else {
    console.log("El valor no es 10 ni 11")
}

// Funciones

function myFunction(): string {
    return "Mi función"
}

console.log(myFunction())

function sumTwoNumbers(firstNumber: number, secondNumber: number): number {
    return firstNumber + secondNumber
}

console.log(sumTwoNumbers(5, 10))

// List

let myList: Array<string> = ["Brais", "Moure", "@mouredev"]
console.log(myList)

// Set

let mySet: Set<string> = new Set(["Brais", "Moure", "@mouredev", "Brais"])
mySet.add("36")
console.log(mySet)

// Mapas

let myMap: Map<string, number> = new Map([["Brais", 36], ["orom_", 40]])
myMap.set("juanitolopez", 23)
console.log(myMap)
console.log(myMap.get("Brais"))

// Bucles

for (const value of myList) {
    console.log(value)
}

let myCounter = 0

while (myCounter < myList.length) {
    console.log(myList[myCounter])
    myCounter++
}

// Clases

class MyClass {
    name: string
    age: number

    constructor(name: string, age: number) {
        this.name = name
        this.age = age
    }
}

let myClass: MyClass = new MyClass("Brais", 36)
console.log(myClass)
console.log(myClass.name)

// Enum
enum MyEnum {
    DART = "dart",
    PYTHON = "python",
    SWIFT = "swift",
    JAVA = "java",
    KOTLIN = "kotlin",
    JAVASCRIPT = "javascript"
}

const myEnum = MyEnum.DART
console.log(myEnum)

File: 08 - Go/hello_world.go
// Día 8: Go
// Clase en vídeo: https://youtu.be/AGiayASyp2Q

// Creación de módulo: go mod init main
package main

import (
	"container/list"
	"fmt"
	"reflect"
)

func main() {

	// Hola mundo

	fmt.Println("Hola, Go!")

	/*
		Esto es un comentario
	*/

	// Variables

	var myString string = "Esto es una cadena de texto"
	fmt.Println(myString)

	myString = "Aquí cambio el valor de la cadena de texto"
	fmt.Println(myString)

	// myString = 6 Error

	var myString2 = "Esto es una cadena de texto"
	fmt.Println(myString2)

	var myInt int = 7
	myInt = myInt + 4
	fmt.Println(myInt)
	fmt.Println(myInt - 1)
	fmt.Println(myInt)

	fmt.Println(myString, myInt)

	fmt.Println(reflect.TypeOf(myInt))

	var myFloat float64 = 6.5
	fmt.Println(myFloat)
	fmt.Println(reflect.TypeOf(myFloat))

	fmt.Println(float64(myInt) + myFloat)

	var myBool bool = false
	myBool = true
	fmt.Println(myBool)

	// Variable declarada e inicializada de manera abreviada

	myString3 := "Esto es una cadena de texto"
	fmt.Println(myString3)

	// Constantes

	const myConst = "Esto es una constante"
	fmt.Println(myConst)

	// Control de flujo

	if myInt == 10 && myString == "Hola" {
		fmt.Println("El valor es 10")
	} else if myInt == 11 || myString == "Hola" {
		fmt.Println("El valor es 11")
	} else {
		fmt.Println("El valor no es 10")
	}

	// Array

	var myArray [3]int
	myArray[0] = 1
	myArray[1] = 2
	myArray[2] = 3
	// myArray[3] = 3 Error
	fmt.Println(myArray[2])
	// fmt.Println(myArray[3]) Error

	// Map

	myMap := make(map[string]int)
	myMap["Brais"] = 36
	myMap["crais01"] = 35
	myMap["hoz98"] = 24
	fmt.Println(myMap)
	fmt.Println(myMap["Brais"])

	myMap2 := map[string]int{"Brais": 36, "crais01": 35, "hoz98": 24}
	fmt.Println(myMap2)

	// List

	myList := list.New()
	myList.PushBack(1)
	myList.PushBack(2)
	myList.PushBack(3)
	fmt.Println(myList.Back().Value)

	// Bucles

	for index := 0; index < len(myArray); index++ {
		fmt.Println(myArray[index])
	}

	for key, value := range myMap {
		fmt.Println(key, value)
	}

	for index, value := range myArray {
		fmt.Println(index, value)
	}

	// Función

	fmt.Println(myFunction())

	// Estructura

	type MyStruct struct {
		name string
		age  int
	}

	myStruct := MyStruct{"Brais", 36}
	fmt.Println(myStruct)
}

func myFunction() string {
	return "Mi función"
}
